### PR TITLE
Fix bugs that occur when two duplicate songs appear on feed from diff…

### PR DESCRIPTION
**Pull request by ChenJesse**
…erent users

#257

### CODE REVIEW

**Comment by loganallen** [comment]
If it works, looks good to me

File: Tempo/Controllers/PlayerTableViewController.swift [hunk 1]
```diff
@@ -47,7 +47,10 @@ class PlayerTableViewController: UIViewController, UITableViewDelegate, UITableV
 			if searchController.isActive {
 				array = filteredPosts
 			}
-            if let row = currentlyPlayingIndexPath?.row, (playerNav.currentPost?.song.equals(other: array[row].song) ?? false && self == (playerNav.playerCell.delegate as! PlayerTableViewController)) {
+            if let row = currentlyPlayingIndexPath?.row,
```
Comment (dennisfedorko) [inline Tempo/Controllers/PlayerTableViewController.swift:50]: you should unwrap currentPost so you don't have to do this ?? nonsense

File: Tempo/Controllers/PlayerTableViewController.swift [hunk 2]
```diff
@@ -47,7 +47,10 @@ class PlayerTableViewController: UIViewController, UITableViewDelegate, UITableV
...
+				(playerNav.currentPost?.song.equals(other: array[row].song) ?? false
+				&& (playerNav.currentPost?.user.equals(other: array[row].user)) ?? false
+				&& self == (playerNav.playerCell.delegate as! PlayerTableViewController)) {
```
Comment (dennisfedorko) [inline Tempo/Controllers/PlayerTableViewController.swift]: dont force cast here, you should optionally cast and unwrap using if let

File: Tempo/Models/User.swift [hunk 3]
```diff
@@ -112,6 +112,10 @@ class User: NSObject, NSCoding {
 		aCoder.encode(updatedAt, forKey: "updated_at")
 		aCoder.encode(username, forKey: "username")
 	}
+	
+	func equals(other: User) -> Bool {
```
Comment (dennisfedorko) [inline Tempo/Models/User.swift:116]: doesn't each user have a unique identifier, isn't that a better than using username?

Comment (ChenJesse) [reply Tempo/Models/User.swift:116]: I don't think so

**Comment by ChenJesse** [comment]
Ok I thought of a slightly better way to organize the checks, I'll get it done by tonight...I just thought we were submitting yesterday so I rushed to bug fix

**Comment by say25** [comment]
SONG ID?

**Comment by ChenJesse** [comment]
What do you mean?

**Comment by say25** [comment]
Songs have a unique Spotify given ID on them. If a user posts truly the same song it by definition has the same song id. that is probably the best way to determine duplicates

File: Tempo/Models/User.swift (hunk 3 above)
Comment (dennisfedorko) [reply Tempo/Models/User.swift:116]: We actually have 3 unique identifiers on a user object: id, fbid, and username

**Comment by ChenJesse** [comment]
Oh, the problem isn't prevent different users from posting the same song, it's dealing with how the logic determines which song is playing when two of them are identical. I do what you're talking about when I do song1.equals(other: song2), which compares IDs. But that alone is not sufficient for when two people you follow post the same song (with same songIDs, since the one further up in the tableview will always get played first)

Dennis was talking about ways to compare User objects, and I didn't see a field that is guaranteed to be unique. Username is guaranteed to be unique, the backend does a check for that whenever you're changing it, which is why I chose it. The only concern is if you change your username, but even if you change the username, the User object is still in memory so it should be fine.

File: Tempo/Models/User.swift (hunk 3 above)
Comment (ChenJesse) [reply Tempo/Models/User.swift:116]: Lol my b, idk how I missed that...

**Comment by dennisfedorko** [comment]
I guess using username is about the same as id, so works for me

**Comment by ChenJesse** [comment]
Ok, fixed up the style, using id for comparing Users, and now defines Post equality as having the same equal users and songs.

Also I found a few other bugs and fixed them:
- Scrolling away from a spinning feedCell and coming back would sometimes stop animation
- Pull to refresh animation continuity of playing song was inconsistent, moved the function call to a better place

**Comment by dennisfedorko** [comment]
Merge it

**Pull request merged**